src/gui_common/dialogs/FilterWindow.cs

```diff
@@ -0,0 +1,256 @@
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Godot;
+
+public class FilterWindow : CustomConfirmationDialog
+{
+    // Paths for nodes
+    [Export]
+    public NodePath FiltersContainerPath = null!;
+
+    private readonly List<Filter> filters = new();
+
+    /// <summary>
+    ///   If redraw is needed.
+    /// </summary>
+    private bool dirty;
+
+    // Nodes
+    private VBoxContainer filtersContainer = null!;
+
+    public override void _Ready()
+    {
+        base._Ready();
+
+        filtersContainer = GetNode<VBoxContainer>(FiltersContainerPath);
+    }
+
+    public override void _Process(float delta)
+    {
+        base._Process(delta);
+
+        if (dirty)
+        {
+            RectSize = GetCombinedMinimumSize();
+        }
+
+        dirty = false;
+    }
+
+    public void Initialize(Filter filter, string defaultText = "--")
+    {
+        ClearFilters();
+        SetupFilter(filter, defaultText);
+    }
+
+    public void ClearFilters()
+    {
+        filters.Clear();
+        filtersContainer.FreeChildren(true);
+    }
+
+    public void SetupFilter(Filter filter, string defaultText = "--")
+    {
+        if (filtersContainer == null)
+            throw new SceneTreeAttachRequired();
+
+        var filterContainer = new HBoxContainer();
+
+        var filterButton = new CustomDropDown();
+        filterButton.Text = defaultText;
+
+        foreach (var option in filter.FilterItemsNames)
+        {
+            filterButton.AddItem(option, false, Colors.White);
+        }
+
+        filterButton.CreateElements();
+        filterButton.Popup.Connect("index_pressed", this, nameof(OnNewFilterCategorySelected),
+            new Godot.Collections.Array { filters.Count });
+
+        filterContainer.AddChild(filterButton);
+        filtersContainer.AddChild(filterContainer);
+        filters.Add(filter);
+
+        dirty = true;
+    }
+
+    private void OnNewFilterCategorySelected(int categoryIndex, int filterIndex)
```

Comment:
I can't get across what you want to do here. Why such a mess? (This func, the previous one and several below)

You can add scenes such as "number slider filter .tscn" or something instead of messing up with casts and multiple throws.

Reply:
![image](https://user-images.githubusercontent.com/60734649/185782559-e91df4e4-c346-45e8-931b-3b9edee69d08.png)